### PR TITLE
Add support for java_annotation_processor rule.

### PR DESCRIPTION
**Pull request by raviagarwal7**
[java_annotation_processor](https://github.com/facebook/buck/blob/master/src/com/facebook/buck/jvm/java/JavaAnnotationProcessorDescription.java) rule allows to share classloaders across different runs for improved build times.

Generates processors rules in `.okbuck/cache/processor/BUCK` file. 

Build rules gets populated with `plugins` parameter instead of `annotation_processors`.

Enable using the experimental extension:
```
okuck {
  experimental {
    useAnnotationProcessorPlugin = true
  }
}
```

Verified that the buck files remains the same before and after the change with `useAnnotationProcessorPlugin` set to false.

### CODE REVIEW

File: buildSrc/src/main/groovy/com/uber/okbuck/OkBuckGradlePlugin.groovy [hunk 1]
```diff
@@ -108,10 +112,20 @@ class OkBuckGradlePlugin implements Plugin<Project> {
             experimentalExtension = experimental
         })
         okBuck.dependsOn(setupOkbuck)
+        okBuck.doLast {
+            // Compose Annotation Processor rules
+            annotationProcessorCache.composeProcessors(project)
+
+            // Finalize Dependency Cache
+            depCache.finalizeDeps()
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/OkBuckGradlePlugin.groovy]: `annotationProcessorCache.finalize()` ?

File: buildSrc/src/main/groovy/com/uber/okbuck/composer/android/AndroidLibraryRuleComposer.groovy [hunk 2]
```diff
@@ -19,7 +19,8 @@ final class AndroidLibraryRuleComposer extends AndroidBuckRuleComposer {
             AndroidLibTarget target,
             List<String> deps,
             final List<String> aidlRuleNames,
-            String appClass) {
+            String appClass,
+            boolean useAAPlugin) {
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/composer/android/AndroidLibraryRuleComposer.groovy]: useApPlugin

File: buildSrc/src/main/groovy/com/uber/okbuck/composer/android/AndroidTestRuleComposer.groovy [hunk 3]
```diff
@@ -20,7 +20,8 @@ final class AndroidTestRuleComposer extends AndroidBuckRuleComposer {
             AndroidLibTarget target,
             List<String> deps,
             final List<String> aidlRuleNames,
-            String appClass) {
+            String appClass,
+            boolean useAPPlugin) {
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/composer/android/AndroidTestRuleComposer.groovy]: useApPlugin

File: buildSrc/src/main/groovy/com/uber/okbuck/composer/android/ExopackageAndroidLibraryRuleComposer.groovy [hunk 4]
```diff
@@ -12,7 +12,7 @@ final class ExopackageAndroidLibraryRuleComposer extends AndroidBuckRuleComposer
         // no instance
     }
 
-    static Rule compose(AndroidAppTarget target) {
+    static Rule compose(AndroidAppTarget target, boolean useAAPlugin) {
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/composer/android/ExopackageAndroidLibraryRuleComposer.groovy]: useApPlugin

File: buildSrc/src/main/groovy/com/uber/okbuck/composer/java/JavaAnnotationProcessorRuleComposer.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+import com.uber.okbuck.template.core.Rule;
+import com.uber.okbuck.template.java.JavaAnnotationProcessorRule;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class JavaAnnotationProcessorRuleComposer extends BuckRuleComposer {
+
+    JavaAnnotationProcessorRuleComposer() {
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/composer/java/JavaAnnotationProcessorRuleComposer.java]: { } on single line?

File: buildSrc/src/main/groovy/com/uber/okbuck/composer/java/JavaAnnotationProcessorRuleComposer.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+                    return new JavaAnnotationProcessorRule()
+                            .processorClasses(i.getAnnotationProcessors())
+                            .name(i.getAnnotationProcessorRuleName())
+                            .deps(depsBuilder.build());
+                })
+                .collect(Collectors.toList());
+    }
+}
+
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/composer/java/JavaAnnotationProcessorRuleComposer.java]: remove extra newline

File: buildSrc/src/main/rocker/com/uber/okbuck/template/jvm/JvmRule.rocker.raw [hunk 7]
```diff
@@ -62,7 +63,14 @@ lib_deps = []
     }
     ],
 }
-@if (valid(annotationProcessors)) {
+@if (valid(annotationProcessors) && useAnnotationProcessorPlugin) {
```
Comment (kageiit) [inline buildSrc/src/main/rocker/com/uber/okbuck/template/jvm/JvmRule.rocker.raw]: It feels a bit weird to leak an experimental flag to the rocker template. This should be business logic contained in the composer

File: buildSrc/src/main/java/com/uber/okbuck/extension/ExperimentalExtension.java [hunk 8]
```diff
@@ -11,7 +11,13 @@
     public boolean transform = false;
 
     /**
-     * Whether to use classpath_abi for lint or not
+     * Whether to use classpath_abi for lint
      */
     public boolean lintWithClasspathAbi = false;
+
+    /**
+     * Whether to use java_annotation_processors rule
```
Comment (kageiit) [inline buildSrc/src/main/java/com/uber/okbuck/extension/ExperimentalExtension.java]: java_annotation_processors -> java_annotation_processor

File: buildSrc/src/main/java/com/uber/okbuck/core/model/jvm/JvmTarget.java [hunk 9]
```diff
@@ -103,25 +147,43 @@ public LintOptions getLintOptions() {
...
+     * List of test annotation processor classes. If annotation processor plugin is enabled
+     * returns the annotation processor plugin rules paths.
      */
     public Set<String> getTestAnnotationProcessors() {
-        return getTestApt().getAnnotationProcessors();
+        ExperimentalExtension experimentalExtension = getOkbuck().getExperimentalExtension();
+        if (experimentalExtension.useAnnotationProcessorPlugin) {
+            return getTestApts()
+                    .stream()
+                    .filter(i -> !i.getAnnotationProcessors().isEmpty())
```
Comment (kageiit) [inline buildSrc/src/main/java/com/uber/okbuck/core/model/jvm/JvmTarget.java]: can we use a better name than `i`

File: buildSrc/src/main/java/com/uber/okbuck/core/model/jvm/JvmTarget.java [hunk 10]
```diff
@@ -103,25 +147,43 @@ public LintOptions getLintOptions() {
...
+     * List of annotation processor classes. If annotation processor plugin is enabled
+     * returns the annotation processor plugin rules paths.
      */
     public Set<String> getAnnotationProcessors() {
-        return getApt().getAnnotationProcessors();
+        ExperimentalExtension experimentalExtension = getOkbuck().getExperimentalExtension();
+        if (experimentalExtension.useAnnotationProcessorPlugin) {
+            return getApts()
+                    .stream()
+                    .filter(i -> !i.getAnnotationProcessors().isEmpty())
```
Comment (kageiit) [inline buildSrc/src/main/java/com/uber/okbuck/core/model/jvm/JvmTarget.java]: can we use a better name than `i`

File: buildSrc/src/main/rocker/com/uber/okbuck/template/jvm/JvmRule.rocker.raw (hunk 7 above)
Comment (raviagarwal7) [reply buildSrc/src/main/rocker/com/uber/okbuck/template/jvm/JvmRule.rocker.raw]: discussed offline, the experimental flag is just a temporary field which we will remove on the next major release after this, so fine to keep it for now. 
Also moving this logic to composer would require modifying bunch of composers instead putting the logic in the rocker template.

File: buildSrc/src/main/java/com/uber/okbuck/extension/ExperimentalExtension.java (hunk 8 above)
Comment (raviagarwal7) [reply buildSrc/src/main/java/com/uber/okbuck/extension/ExperimentalExtension.java]: done

File: buildSrc/src/main/java/com/uber/okbuck/core/model/jvm/JvmTarget.java (hunk 9 above)
Comment (raviagarwal7) [reply buildSrc/src/main/java/com/uber/okbuck/core/model/jvm/JvmTarget.java]: done

File: buildSrc/src/main/java/com/uber/okbuck/core/model/jvm/JvmTarget.java (hunk 10 above)
Comment (raviagarwal7) [reply buildSrc/src/main/java/com/uber/okbuck/core/model/jvm/JvmTarget.java]: done

File: buildSrc/src/main/groovy/com/uber/okbuck/composer/java/JavaAnnotationProcessorRuleComposer.java (hunk 6 above)
Comment (raviagarwal7) [reply buildSrc/src/main/groovy/com/uber/okbuck/composer/java/JavaAnnotationProcessorRuleComposer.java]: done

File: buildSrc/src/main/groovy/com/uber/okbuck/composer/java/JavaAnnotationProcessorRuleComposer.java (hunk 5 above)
Comment (raviagarwal7) [reply buildSrc/src/main/groovy/com/uber/okbuck/composer/java/JavaAnnotationProcessorRuleComposer.java]: done

File: buildSrc/src/main/groovy/com/uber/okbuck/composer/android/ExopackageAndroidLibraryRuleComposer.groovy (hunk 4 above)
Comment (raviagarwal7) [reply buildSrc/src/main/groovy/com/uber/okbuck/composer/android/ExopackageAndroidLibraryRuleComposer.groovy]: done

File: buildSrc/src/main/groovy/com/uber/okbuck/composer/android/AndroidTestRuleComposer.groovy (hunk 3 above)
Comment (raviagarwal7) [reply buildSrc/src/main/groovy/com/uber/okbuck/composer/android/AndroidTestRuleComposer.groovy]: done

File: buildSrc/src/main/groovy/com/uber/okbuck/composer/android/AndroidLibraryRuleComposer.groovy (hunk 2 above)
Comment (raviagarwal7) [reply buildSrc/src/main/groovy/com/uber/okbuck/composer/android/AndroidLibraryRuleComposer.groovy]: done

File: buildSrc/src/main/groovy/com/uber/okbuck/OkBuckGradlePlugin.groovy (hunk 1 above)
Comment (raviagarwal7) [reply buildSrc/src/main/groovy/com/uber/okbuck/OkBuckGradlePlugin.groovy]: done

File: buildSrc/src/main/groovy/com/uber/okbuck/OkBuckGradlePlugin.groovy [hunk 11]
```diff
@@ -160,7 +170,7 @@ class OkBuckGradlePlugin implements Plugin<Project> {
                 }
 
                 // Fetch transform deps if needed
-                if (experimental.transform) {
+                if (okbuckExt.experimentalExtension.transform) {
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/OkBuckGradlePlugin.groovy]: do we need this?

File: buildSrc/src/main/groovy/com/uber/okbuck/composer/java/JavaAnnotationProcessorRuleComposer.java [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class JavaAnnotationProcessorRuleComposer extends BuckRuleComposer {
+
+    JavaAnnotationProcessorRuleComposer() { }
+
+    public static List<Rule> compose(Collection<Scope> scopeList) {
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/composer/java/JavaAnnotationProcessorRuleComposer.java]: lets add javadocs here

File: buildSrc/src/main/groovy/com/uber/okbuck/core/dependency/DependencyCache.groovy [hunk 13]
```diff
@@ -180,24 +183,55 @@ class DependencyCache {
...
+        String key = dependency.cacheName
+        String extensions = processorExtensions.get(key)
+
+        if (extensions == null) {
+            extensions = getJarFileContent(dependency.depFile,
+                    "META-INF/services/com.google.auto.value.extension.AutoValueExtension")
+            processorExtensions.set(key, extensions)
+        }
+
+        return extensions && extensions.length() > 0
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/dependency/DependencyCache.groovy]: can we do null check here

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidInstrumentationTarget.groovy [hunk 14]
```diff
@@ -13,9 +17,27 @@ abstract class AndroidInstrumentationTarget extends AndroidAppTarget {
     }
 
     // TODO: Update to use variant once issue solved: https://youtrack.jetbrains.com/issue/KT-23411
+    @Override
+    List<Scope> getApts() {
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidInstrumentationTarget.groovy]: can we call it getAptScopes

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidInstrumentationTarget.groovy [hunk 15]
```diff
@@ -13,9 +17,27 @@ abstract class AndroidInstrumentationTarget extends AndroidAppTarget {
...
     @Override
     Scope getApt() {
-        return Scope.from(project, isKapt ? "kaptAndroidTest" : baseVariant.annotationProcessorConfiguration)
+        def configuration = isKapt ? "kaptAndroidTest" : baseVariant.annotationProcessorConfiguration
+
+        ExperimentalExtension experimentalExtension = getOkbuck().getExperimentalExtension();
+        if (experimentalExtension.useAnnotationProcessorPlugin) {
+            if (!ProjectUtil.getAnnotationProcessorCache(getProject())
+                    .hasEmptyAnnotationProcessors(getProject(), configuration)) {
+                return Scope.from(getProject(), (Configuration) null);
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidInstrumentationTarget.groovy]: lets create EMPTY scope in Scope as static

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidInstrumentationTarget.groovy [hunk 16]
```diff
@@ -13,9 +17,27 @@ abstract class AndroidInstrumentationTarget extends AndroidAppTarget {
...
+                isKapt ? "kaptAndroidTest" : baseVariant.annotationProcessorConfiguration)
+    }
+
+
     @Override
     Scope getApt() {
-        return Scope.from(project, isKapt ? "kaptAndroidTest" : baseVariant.annotationProcessorConfiguration)
+        def configuration = isKapt ? "kaptAndroidTest" : baseVariant.annotationProcessorConfiguration
+
+        ExperimentalExtension experimentalExtension = getOkbuck().getExperimentalExtension();
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidInstrumentationTarget.groovy]: can we DRY this

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy [hunk 17]
```diff
@@ -154,16 +157,51 @@ abstract class AndroidTarget extends JvmTarget {
...
                         unitTestVariant ? unitTestVariant.annotationProcessorConfiguration : null)
+
+    }
+
+    @Override
+    Scope getApt() {
+        def configuration = isKapt ?
+                "kapt${baseVariant.name.capitalize()}" : baseVariant.annotationProcessorConfiguration
+
+        ExperimentalExtension experimentalExtension = getOkbuck().getExperimentalExtension();
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy]: DRY

**Comment by kageiit** [comment]
Can you please rebase as well?

File: buildSrc/src/main/groovy/com/uber/okbuck/core/dependency/DependencyCache.groovy (hunk 13 above)
Comment (raviagarwal7) [reply buildSrc/src/main/groovy/com/uber/okbuck/core/dependency/DependencyCache.groovy]: done

File: buildSrc/src/main/java/com/uber/okbuck/core/model/base/Scope.java [hunk 18]
```diff
@@ -211,49 +217,91 @@ public static Scope from(Project project, Configuration configuration) {
...
-                            return target.getProp(okBuckExtension.annotationProcessors, ImmutableList.of());
-                        })
-                        .flatMap(List::stream))
-                .filter(StringUtils::isNotEmpty)
-                .collect(Collectors.toSet());
+    /**
+     * Returns the UUID for the annotation processors of the scope.
+     * @return String UUID
+     */
+    public String getAnnotationProcessorsUUID() {
```
[the remark this answers is not in the document]
Comment (raviagarwal7) [reply buildSrc/src/main/java/com/uber/okbuck/core/model/base/Scope.java]: done

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/AnnotationProcessorCache.java [hunk 19]
```diff
@@ -0,0 +1,202 @@
...
+        Map<Set<Dependency>, Scope> depToScope = composeProcessors(
+                project, configuration.getAllDependencies(), false);
+
+        return depToScope
+                .values()
+                .stream()
+                .anyMatch(scope -> scope.getAnnotationProcessors().isEmpty());
+    }
+
+    private Map<Set<Dependency>, Scope> composeProcessors(Project project,
```
[the remark this answers is not in the document]
Comment (raviagarwal7) [reply buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/AnnotationProcessorCache.java]: done

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/AnnotationProcessorCache.java [hunk 20]
```diff
@@ -0,0 +1,202 @@
...
+                    String.format("No valid configuration found for '%s' in project '%s'",
+                            configurationString, project.getDisplayName()));
+        }
+        return configuration;
+    }
+
+    /**
+     * Write the buck file for the java_annotation_processor rules.
+     */
+    public void finalizeProcessors() {
```
[the remark this answers is not in the document]
Comment (raviagarwal7) [reply buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/AnnotationProcessorCache.java]: done

**Pull request merged**